components/esp_dns/esp_dns.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * SPDX-FileCopyrightText: 2025 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+/**
+ * esp_dns.c - Custom DNS module for ESP32 with multiple protocol support
+ *
+ * Provides DNS resolution capabilities with support for various protocols:
+ * - Standard UDP/TCP DNS (Port 53)
+ * - DNS over TLS (DoT) (Port 853)
+ * - DNS over HTTPS (DoH) (Port 443)
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "freertos/FreeRTOS.h"
+#include "freertos/semphr.h"
+#include "freertos/task.h"
+#include "esp_log.h"
+//#include "esp_netif.h"
+#include "esp_system.h"
+#include "lwip/api.h"
+#include "lwip/dns.h"
+#include "lwip/sockets.h"
+#include "lwip/netdb.h"
+
+#include "esp_dns.h"
+#include "esp_dns_priv.h"
+#include "esp_dns_udp.h"
+#include "esp_dns_tcp.h"
+#include "esp_dns_dot.h"
+#include "esp_dns_doh.h"
+
+#define TAG "ESP_DNS"
+
+
+/* Global DNS handle instance */
+static esp_dns_handle_t g_dns_handle = NULL;
```

Comment:
This could use a similar approach as we use in the `tcp_transport` and create a resolver list. Then the user could enable as many resolvers as needed and select the current resolver in place. This would also open the space for custom resolver. 

We could set this improvement for the next version of the component.